File: 백준/Silver/11286. 절댓값 힙/절댓값 힙.py
#  **************************************************************************  #
#                                                                              #
#                                                       :::    :::    :::      #
#    Problem Number: 11286                             :+:    :+:      :+:     #
#                                                     +:+    +:+        +:+    #
#    By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+   #
#                                                   +#+      +#+        +#+    #
#    https://boj.kr/11286                          #+#        #+#      #+#     #
#    Solved: 2025/08/19 02:14:51 by hrkim2001     ###          ###   ##.kr     #
#                                                                              #
#  **************************************************************************  #
from queue import PriorityQueue
import sys
print = sys.stdout.write
input = sys.stdin.readline

N = int(input())
pq = PriorityQueue()

for i in range(N):
    x = int(input())
    if x == 0:
        if pq.empty():
            print('0\n')
        else:
            print(str(pq.get()[1]) + '\n')
    if x != 0:
        pq.put((abs(x), x))

File: 백준/Silver/11286. 절댓값 힙/절댓값 힙.java
/* ************************************************************************** */
/*                                                                            */
/*                                                      :::    :::    :::     */
/*   Problem Number: 11286                             :+:    :+:      :+:    */
/*                                                    +:+    +:+        +:+   */
/*   By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+  */
/*                                                  +#+      +#+        +#+   */
/*   https://boj.kr/11286                          #+#        #+#      #+#    */
/*   Solved: 2026/03/08 23:47:56 by hrkim2001     ###          ###   ##.kr    */
/*                                                                            */
/* ************************************************************************** */
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Integer> Q = new PriorityQueue<>((o1, o2) -> {
            if(Math.abs(o1) == Math.abs(o2)) {
                return o1 < o2 ? -1 : 1;
            } else {
                return Math.abs(o1) < Math.abs(o2) ? -1 : 1;
            }
        });

        for(int i = 0; i < N; i++) {
            int num = Integer.parseInt(br.readLine());
            if(num == 0) {
                if(Q.isEmpty()) bw.write("0\n");
                else bw.write(Q.poll() + "\n");
            } else  {
                Q.add(num);
            }
        }
        
        bw.flush();
        bw.close();
        br.close();
    }
}